modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java

```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
+                step1 = state.targetGridGeometry.getGridToCRS(CORNER);
             } else {
-                step1 = targetGG.getGridToCRS(CORNER);
-                if (!targetGG.isDefined(GridGeometry2D.GRID_RANGE_BITMASK)) {
+                step1 = state.targetGridGeometry.getGridToCRS(CORNER);
+                if (!state.targetGridGeometry.isDefined(GridGeometry2D.GRID_RANGE_BITMASK)) {
                     GeneralEnvelope gridRange = CRS.transform(step1.inverse(), targetEnvelope);
                     // According OpenGIS specification, GridGeometry maps pixel's center.
-                    targetGG =
+                    state.targetGridGeometry =
                             new GridGeometry2D(
                                     new GeneralGridEnvelope(gridRange, PixelInCell.CELL_CENTER),
                                     step1,
-                                    targetCRS);
+                                    state.targetCRS);
                 }
             }
             /*
              * Computes the final transform.
              */
             allSteps =
-                    mtFactory.createConcatenatedTransform(
-                            mtFactory.createConcatenatedTransform(step1, step2), step3);
+                    state.mtFactory.createConcatenatedTransform(
+                            state.mtFactory.createConcatenatedTransform(step1, step2), step3);
         }
-        final MathTransform allSteps2D = toMathTransform2D(allSteps, mtFactory, targetGG);
+        final MathTransform allSteps2D =
+                toMathTransform2D(allSteps, state.mtFactory, state.targetGridGeometry);
         if (!(allSteps2D instanceof MathTransform2D)) {
             // Should not happen with Geotools implementations. May happen
             // with some external implementations, but should stay unusual.
             throw new TransformException(Errors.format(ErrorKeys.NO_TRANSFORM2D_AVAILABLE));
         }
+        state.transformSteps =
+                new ResampleState.TransformSteps(
+                        step1, step2, step3, allSteps, (MathTransform2D) allSteps2D);
+    }
 
-        ////////////////////////////////////////////////////////////////////////////////////////
-        ////                                                                                ////
-        ////            STEP 1: Extracts needed informations from the parameters            ////
-        ////            STEP 2: Creates the "target to source" MathTransform                ////
-        //// =======>>  STEP 3: Computes the target image layout                   <<====== ////
-        ////            STEP 4: Applies the JAI operation ("Affine", "Warp", etc)           ////
-        ////                                                                                ////
-        ////////////////////////////////////////////////////////////////////////////////////////
+    private static void createSourcePlanarImage(ResampleState state) {
```

Comment:
I don't see expensive code here, could have been part of the state constructor too.